### PR TITLE
[MXNET-1209] Tutorial transpose reshape 

### PR DESCRIPTION
**Pull request by NRauschmayr**
## Description ##
Adding a tutorial that explains the difference between reshape and transpose operators
@ThomasDelteil @Ishitori can you please have a look? Thanks!

http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-13208/3/tutorials/basic/reshape_transpose.html

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)

### CODE REVIEW

**Comment by anirudhacharya** [comment]
@mxnet-label-bot add [pr-awaiting-review]

File: docs/tutorials/basic/reshape_transpose.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+```
+
+(1L, 3L, 64L, 2L, 64L, 2L) <!--notebook-skip-line-->
+
+(1L, 3L, 128L, 128L) <!--notebook-skip-line-->
+
+This was just a toy example. But such transformations are for instance done in image superresolution where you increase width and height of the input image and ```x``` would be the output of a CNN that computes an upscale feature vector.
+
+#### Check out the MXNet documentation for more details
+http://mxnet.incubator.apache.org/test/api/python/ndarray.html#mxnet.ndarray.NDArray.reshape
```
Comment (anirudhacharya) [inline docs/tutorials/basic/reshape_transpose.md]: This shows as just text, it should be a clickable link. 

But a bigger point is I don't think we need this section at all. Can these two documentation links be added as a hyperlink to the first occurrence of the terms "Reshape" and "Transpose" at the beginning of the tutorial and remove this section.

You are saying that, "go to the the documentation for more details" but I think this tutorial contains more details than the documentation. :)

can you also please update your PR description with this link - http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-13208/3/tutorials/basic/reshape_transpose.html. It will help reviewers, especially when reviewing tutorials.

**Comment by anirudhacharya** [comment]
There are a bunch of links pointing to a personal repo instead of dmlc/web-data. Please replace those.

The rest LGTM.

**Comment by simoncorstonoliver** [comment]
It would be helpful for users to include some discussion of the common errors that they see e.g. when there's a tensor shape mismatch between layers of a NN. If we included the actual error message then people would find this tutorial when they Googled for the error message. This would segue nicely to the example where you show how you can't just perform these operations to make the error go away; you have to actually know what you're doing.

For users who already know what tensor reshaping is, the information they need (the "how" rather than the "why") is in a few non-obvious places. Maybe add something near the top with pointers to ndarray docs for more comprehensive documentation: https://mxnet.incubator.apache.org/tutorials/basic/ndarray.html

**Comment by NRauschmayr** [comment]
I added more examples e.g. common pitfalls and errors 
I also changed a couple of minor things.

File: docs/tutorials/basic/reshape_transpose.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
+
+## Difference between reshape and transpose operators
+Modyfing the shape of tensors is a very common operation in Deep Learning. For instance, when using pretrained neural networks it is often required to adjust input data dimensions to correspond to what the network has been trained on, e.g. tensors of shape `[batch_size, channels, width, height]`.  This notebook discusses briefly the difference between the operators [Reshape](http://mxnet.incubator.apache.org/test/api/python/ndarray.html#mxnet.ndarray.NDArray.reshape) and [Transpose](http://mxnet.incubator.apache.org/test/api/python/ndarray.html#mxnet.ndarray.transpose). Both allow to change the shape, however they are not the same and are commonly mistaken.
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: Modyfing -> Modifying

often required to adjust input data dimension --> often necessary to adjust the input data dimension 

Both allow to --> Both allow you to

File: docs/tutorials/basic/reshape_transpose.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+![png](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/basic/transpose_reshape/cat.png) <!--notebook-skip-line-->
+
+
+The color image has the following properties:
+* width: 210 pixels
+* height: 200 pixels
+* colors: 3 (RGB)
+
+Now lets reshape the image in order to exchange width and height dimension.
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: width and height dimensions

File: docs/tutorials/basic/reshape_transpose.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+print (reshaped.shape)
+plt.imshow(reshaped)
+plt.axis("off")
+```
+(200, 210, 3)<!--notebook-skip-line-->
+
+![png](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/basic/transpose_reshape/reshaped_image.png) <!--notebook-skip-line-->
+
+
+As we can see the first and second dimensions have changed. However the image can't be identified as cat anylonger. In order to understand what happened, let's have a look at the image below.
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: any longer

File: docs/tutorials/basic/reshape_transpose.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+```
+
+![png](https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/basic/transpose_reshape/transposed_image.png) <!--notebook-skip-line-->
+
+
+As we can see width and height changed, by rotating pixel values by 90 degrees. Transpose does the following:
+
+<img src="https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/doc/tutorials/basic/transpose_reshape/transpose.png" style="width:700px;height:300px;">
+
+As shown in the diagram, the axis have been flipped: pixel values that have been in the first row are now in the first column.
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: the axis have --> the axes have

File: docs/tutorials/basic/reshape_transpose.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+## When to transpose/reshape with MXNet
+In this chapter we discuss when transpose and reshape is used in MXNet. 
+#### Channel first for images
+Images are usually stored in the format height, wight, channel. When working with [convolutional](https://mxnet.incubator.apache.org/api/python/gluon/nn.html#mxnet.gluon.nn.Conv1D) layers, MXNet expects the layout to be `NCHW` (batch, channel, height, width). MXNet uses this layout because of performance reasons on the GPU. Consequently, images need to be transposed to have the right format. For instance, you may have a function like the following:
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: worth calling out that this channel ordering is different from TF?

File: docs/tutorials/basic/reshape_transpose.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+network.hybridize()
+network.initialize(mx.init.Xavier(), ctx=mx.cpu())
+a = mx.random.uniform(shape=(1,100,1000))
+network(a)
+output = network(a)
+print (output.shape)
+```
+(1, 999, 128) <!--notebook-skip-line-->
+#### Advanced reshaping with MXNet ndarrays
+It is sometimes useful to automatically infer the shape of tensors. Especially when you deal with very deep neural networks, it may not always be clear what the shape of a tensor is after a specific layer. For instance you may want the tensor to be two-dimensional where one dimension is the known batch_size. With ```mx.nd.array(-1, batch_size)``` the first dimension will be automatically inferred. Here a simplified example:
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: Here a simplified --> Here is a simplified

File: docs/tutorials/basic/reshape_transpose.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+/home/ubuntu/anaconda3/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so(_ctypes_callproc+0x2ce)[0x7fdf46547e2e]
+/home/ubuntu/anaconda3/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so(+0x12865)[0x7fdf46548865]
+python(_PyObject_FastCallDict+0x8b)[0x56457eba2d7b]
+python(+0x19e7ce)[0x56457ec327ce]
+python(_PyEval_EvalFrameDefault+0x2fa)[0x56457ec54cba]
+python(+0x197dae)[0x56457ec2bdae]
+[...]
+
+```
+This is happening when you your data does not have the shape ```[batch_size, channel, width, height]``` e.g. your data may be a one-dimensional vector or when the color channel may be the last dimension instead of the second one.
```
Comment (simoncorstonoliver) [inline docs/tutorials/basic/reshape_transpose.md]: This is happening --> This happens

**Comment by vandanavk** [comment]
LGTM.

@NRauschmayr Could you trigger the CI again?

@simoncorstonoliver @ThomasDelteil for another round of review

File: docs/tutorials/basic/reshape_transpose.md [hunk 9]
```diff
@@ -1,7 +1,6 @@
 ## Difference between reshape and transpose operators
 
 What does it mean if MXNet gives you an error like the this?
-
```
Comment (piyushghai) [inline docs/tutorials/basic/reshape_transpose.md]: Hi @NRauschmayr , There's also another way of pushing an empty commit in case you need to re-trigger the CI. 
You can try ```git commit --allow-empty -m "<commit-message>"``` :)

**Comment by ThomasDelteil** [comment]
@NRauschmayr checking with @marcoabreu  why the build status was not propagated back to the PR status check, which is necessary for merging

**Comment by marcoabreu** [comment]
http://jenkins.mxnet-ci.amazon-ml.com/job/mxnet-validation/job/windows-gpu/view/change-requests/job/PR-13208/4/console

**Comment by NRauschmayr** [comment]
Any update on this?
